src/mono/mono/utils/mono-time.c

```diff
@@ -229,3 +233,48 @@ mono_100ns_datetime_from_timeval (struct timeval tv)
 
 #endif
 
+#ifdef HOST_DARWIN
+
+static clock_serv_t sampling_clock_service;
+
+guint64
+mono_clock_get_time_ns (void)
+{
+	kern_return_t ret;
+	mach_timespec_t mach_ts;
+
+	do {
+		ret = clock_get_time (sampling_clock_service, &mach_ts);
+	} while (ret == KERN_ABORTED);
+
+	if (ret != KERN_SUCCESS)
+		g_error ("%s: clock_get_time () returned %d", __func__, ret);
+
+	return ((guint64) mach_ts.tv_sec * 1000000000) + (guint64) mach_ts.tv_nsec;
+}
+
+#elif defined(HOST_LINUX)
+
+static clockid_t sampling_posix_clock;
```

Comment:
Same: the POSIX version of `clock_init()` would initialize this, but it's now always zero.